libs/platform/src/lib/components/approval-flow/approval-flow.component.spec.ts

```diff
@@ -0,0 +1,307 @@
+        imgUrl: 'https://randomuser.me/api/portraits/women/87.jpg'
+    },
+    {
+        id: 'uid66151',
+        name: 'Sofia Hanson',
+        description: 'Accounting team',
+        imgUrl: 'https://randomuser.me/api/portraits/women/24.jpg'
+    },
+    {
+        id: 'uid66161',
+        name: 'Jill Fuller',
+        description: 'Accounting team',
+        imgUrl: 'https://randomuser.me/api/portraits/women/64.jpg'
+    },
+    {
+        id: 'uid66171',
+        name: 'Ella Franklin',
+        description: 'Accounting team',
+        imgUrl: 'https://randomuser.me/api/portraits/women/55.jpg'
+    }
+];
+
+const simpleGraph: ApprovalProcess = {
+    watchers: [getRandomUser(), getRandomUser(), getRandomUser()],
+    nodes: [
+        {
+            id: 'ID1',
+            name: 'node name',
+            description: 'node description',
+            approvers: [getRandomUser()],
+            status: 'approved',
+            targets: ['ID2'],
+            dueDate: new Date(),
+            createDate: new Date()
+        },
+        {
+            id: 'ID2',
+            name: 'node name',
+            description: 'node description',
+            approvers: [getRandomUser()],
+            status: 'in progress',
+            targets: ['ID3'],
+            dueDate: new Date(),
+            createDate: new Date()
+        },
+        {
+            id: 'ID3',
+            name: 'node name',
+            description: 'node description',
+            approvers: [getRandomUser()],
+            status: 'in progress',
+            targets: ['ID4'],
+            dueDate: new Date(),
+            createDate: new Date()
+        },
+        {
+            id: 'ID4',
+            name: 'node name',
+            description: 'node description',
+            approvers: [getRandomUser()],
+            status: 'not started',
+            targets: [],
+            dueDate: new Date(),
+            createDate: new Date()
+        }
+    ]
+};
+
+function getRandomUser(): ApprovalUser {
+    return users[Math.floor(Math.random() * users.length)];
+}
+
+export class TestApprovalFlowDataSource implements ApprovalDataSource {
+    fetch(): Observable<ApprovalProcess> {
+        return of(simpleGraph);
+    }
+    fetchUser(id: string): Observable<any> {
+        const user = users.find(u => u.id === id);
+
+        return of({
+            phone: Math.random().toFixed(13).toString().replace('0.', ''),
+            email: `${user.name.toLowerCase().split(' ').join('.')}@company.com`
+        });
+    }
+    updateWatchers(watchers: ApprovalUser[]): void {}
+    updateApproval(approval: ApprovalNode): void {}
+    updateApprovals(approvals: ApprovalNode[]): void {}
+    sendReminders(members: ApprovalUser[], approval: ApprovalNode): void {}
+}
+
+const TEST_APPROVAL_FLOW_TITLE = 'Test title';
+
+@Component({
+    selector: 'fdp-test-approval-flow',
+    template: `
+        <fdp-approval-flow [title]="title" [dataSource]="dataSource"></fdp-approval-flow>`
```

Comment:
Just a query, how datasource is helping here us comparing to using angular basic approaches(*ngFor, array etc)

Reply:
Using `dataSource` was a requirement from the tech spec.

I don't get why `*ngFor, array` is compared to the DataSource here, they work independently, DataSource provides methods for retrieving approval flow's data and manipulating that data in any way, *ngFor and etc are internal Angular tools that are used quite a lot in this component's code.

As far as I can tell, for a complex component like this one using DataSource is preferable way, otherwise the component's number of inputs and outputs would be quite big. Also using DataSource allows to have all `data` logic in one place